Fix complex printing for sci_mode=True

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40513 Fix complex printing for sci_mode=True**

This PR makes the following changes:
1. Complex Printing now uses print formatting for it's real and imaginary values and they are joined at the end.
2. Adding 1. naturally fixes the printing of complex tensors in sci_mode=True

```
>>> torch.tensor(float('inf')+float('inf')*1j)
tensor(nan+infj)
>>> torch.randn(2000, dtype=torch.cfloat)
tensor([ 0.3015-0.2502j, -1.1102+1.2218j, -0.6324+0.0640j,  ...,
        -1.0200-0.2302j,  0.6511-0.1889j, -0.1069+0.1702j])
>>> torch.tensor([1e-3, 3+4j, 1e-5j, 1e-2+3j, 5+1e-6j])
tensor([1.0000e-03+0.0000e+00j, 3.0000e+00+4.0000e+00j, 0.0000e+00+1.0000e-05j,
        1.0000e-02+3.0000e+00j, 5.0000e+00+1.0000e-06j])
>>> torch.randn(3, dtype=torch.cfloat)
tensor([ 1.0992-0.4459j,  1.1073+0.1202j, -0.2177-0.6342j])
>>> x = torch.tensor([1e2, 1e-2])
>>> torch.set_printoptions(sci_mode=False)
>>> x
tensor([  100.0000,     0.0100])
>>> x = torch.tensor([1e2, 1e-2j])
>>> x
tensor([100.+0.0000j,   0.+0.0100j])
```

fixes https://github.com/pytorch/pytorch/issues/38285

Differential Revision: [D22309294](https://our.internmc.facebook.com/intern/diff/D22309294)

anjali411 committed:
Fix complex printing for sci_mode=True

[ghstack-poisoned]

anjali411 committed:
Update on "Fix complex printing for sci_mode=True"


This PR makes the following changes:
1. Complex Printing now uses print formatting for it's real and imaginary values and they are joined at the end.
2. Adding 1. naturally fixes the printing of complex tensors in sci_mode=True

```
>>> torch.tensor(float('inf')+float('inf')*1j)
tensor(nan+infj)
>>> torch.randn(2000, dtype=torch.cfloat)
tensor([ 0.3015-0.2502j, -1.1102+1.2218j, -0.6324+0.0640j,  ...,
        -1.0200-0.2302j,  0.6511-0.1889j, -0.1069+0.1702j])
>>> torch.tensor([1e-3, 3+4j, 1e-5j, 1e-2+3j, 5+1e-6j])
tensor([1.0000e-03+0.0000e+00j, 3.0000e+00+4.0000e+00j, 0.0000e+00+1.0000e-05j,
        1.0000e-02+3.0000e+00j, 5.0000e+00+1.0000e-06j])
>>> torch.randn(3, dtype=torch.cfloat)
tensor([ 1.0992-0.4459j,  1.1073+0.1202j, -0.2177-0.6342j])
>>> x = torch.tensor([1e2, 1e-2])
>>> torch.set_printoptions(sci_mode=False)
>>> x
tensor([  100.0000,     0.0100])
>>> x = torch.tensor([1e2, 1e-2j])
>>> x
tensor([100.+0.0000j,   0.+0.0100j])
```

[ghstack-poisoned]

ezyang commented:
waiting on tests before review

anjali411 committed:
Update on "Fix complex printing for sci_mode=True"


This PR makes the following changes:
1. Complex Printing now uses print formatting for it's real and imaginary values and they are joined at the end.
2. Adding 1. naturally fixes the printing of complex tensors in sci_mode=True

```
>>> torch.tensor(float('inf')+float('inf')*1j)
tensor(nan+infj)
>>> torch.randn(2000, dtype=torch.cfloat)
tensor([ 0.3015-0.2502j, -1.1102+1.2218j, -0.6324+0.0640j,  ...,
        -1.0200-0.2302j,  0.6511-0.1889j, -0.1069+0.1702j])
>>> torch.tensor([1e-3, 3+4j, 1e-5j, 1e-2+3j, 5+1e-6j])
tensor([1.0000e-03+0.0000e+00j, 3.0000e+00+4.0000e+00j, 0.0000e+00+1.0000e-05j,
        1.0000e-02+3.0000e+00j, 5.0000e+00+1.0000e-06j])
>>> torch.randn(3, dtype=torch.cfloat)
tensor([ 1.0992-0.4459j,  1.1073+0.1202j, -0.2177-0.6342j])
>>> x = torch.tensor([1e2, 1e-2])
>>> torch.set_printoptions(sci_mode=False)
>>> x
tensor([  100.0000,     0.0100])
>>> x = torch.tensor([1e2, 1e-2j])
>>> x
tensor([100.+0.0000j,   0.+0.0100j])
```
TODO: 
Add tests

[ghstack-poisoned]

anjali411 commented:
> waiting on tests before review

@ezyang added tests!

mruberry commented on test/test_torch.py:
```diff
@@ -3864,9 +3864,16 @@ def test_print(self):
                 str(obj)
 
             # test complex tensor
+            # complex tensor print uses two formatters, one for real values
```
Meta question: these tests seem OK, but would it be possible to compare more directly with NumPy's printing? 

The only differences should be that NumPy writes "array" instead of "tensor" and will write the dtype at difference times since both PyTorch and NumPy only write a tensor's dtype if it's not equal to the default float scalar type. Right? Seems like this test could account for those differences.

anjali411 replied:
There are other differences between numpy array and torch tensor printing like extra 0s, when we choose to have sci_mode etc, which is prob why other dtypes currently don't compare torch out with numpy out

mruberry commented on torch/_tensor_str.py:
```diff
@@ -228,9 +258,14 @@ def _tensor_str(self, indent):
 
+    if self.dtype.is_complex:
+        real_formatter = _Formatter(get_summarized_data(self.real) if summarize else self.real)
+        imag_formatter = _Formatter(get_summarized_data(self.imag) if summarize else self.imag)
```
How do the tests trigger summarize True/False?

If summarize is False then real_formatter/imag_formatter = self.real/self.imag. These are tensors which are passed to _complex_tensor_str_with_formatter(), which can call `_scalar_str(self.real, real_formatter)`, which seems like a nonsense call when real_formatter is a tensor?

anjali411 replied:
I am not sure what you are talking about. `real_formatter` and `imag_formatter` are formatters at all places.

anjali411 replied:
also I have already added the test for summarize as mentioned above :D

anjali411 replied:
resolved https://github.com/pytorch/pytorch/pull/40513#discussion_r447752639

mruberry commented on torch/_tensor_str.py:
```diff
@@ -212,6 +191,57 @@ def _tensor_str_with_formatter(self, indent, formatter, summarize):
+    if dim == 1:
+        return _complex_vector_str(self, indent, real_formatter, imag_formatter, summarize)
+
+    if summarize and self.size(0) > 2 * PRINT_OPTS.edgeitems:
```
I worry about the redundancy (and maintainability) of this function vs. the existing `_tensor_str_with_formatter`. In fact, even `_vector_str ` and `_complex_vector_str` have a lot of similarities. Would it be possible to push the complex code into `_vector_str`, or to make `_tensor_str_with_formatter` polymorphic to avoid some of the code duplication?

anjali411 replied:
yeah you are right. I was thinking about it too. I updated the PR to reuse the existing code wherever there was a common logic between real and complex printing

mruberry replied:
You're correct - my mistake, I misread the line where they're constructed.

anjali411 committed:
Update on "Fix complex printing for sci_mode=True"


This PR makes the following changes:
1. Complex Printing now uses print formatting for it's real and imaginary values and they are joined at the end.
2. Adding 1. naturally fixes the printing of complex tensors in sci_mode=True

```
>>> torch.tensor(float('inf')+float('inf')*1j)
tensor(nan+infj)
>>> torch.randn(2000, dtype=torch.cfloat)
tensor([ 0.3015-0.2502j, -1.1102+1.2218j, -0.6324+0.0640j,  ...,
        -1.0200-0.2302j,  0.6511-0.1889j, -0.1069+0.1702j])
>>> torch.tensor([1e-3, 3+4j, 1e-5j, 1e-2+3j, 5+1e-6j])
tensor([1.0000e-03+0.0000e+00j, 3.0000e+00+4.0000e+00j, 0.0000e+00+1.0000e-05j,
        1.0000e-02+3.0000e+00j, 5.0000e+00+1.0000e-06j])
>>> torch.randn(3, dtype=torch.cfloat)
tensor([ 1.0992-0.4459j,  1.1073+0.1202j, -0.2177-0.6342j])
>>> x = torch.tensor([1e2, 1e-2])
>>> torch.set_printoptions(sci_mode=False)
>>> x
tensor([  100.0000,     0.0100])
>>> x = torch.tensor([1e2, 1e-2j])
>>> x
tensor([100.+0.0000j,   0.+0.0100j])
```

[ghstack-poisoned]

anjali411 commented on test/test_torch.py:
```diff
@@ -3990,6 +4004,32 @@ def test_print(self):
+            y = x.as_strided(size=(100, 2, 10), stride=(2 * 2 * 10, 2 * 10, 1))
+            self.assertEqual(str(y), y.__repr__())
+            expected_str = '''\
+tensor([[[1.+1.j, 1.+1.j, 1.+1.j,  ..., 1.+1.j, 1.+1.j, 1.+1.j],
```
@mruberry here's the test for summarize. if a tensor has > 1000 elements, summarize is automatically triggered.

mruberry replied:
Cool - thanks for pointing that out.

anjali411 commented on test/test_torch.py:
```diff
@@ -3903,6 +3910,13 @@ def test_print(self):
             self.assertEqual(x.__repr__(), str(x))
             self.assertExpectedInline(str(x), '''tensor([4.0000,    inf, 1.5000,   -inf, 0.0000,    nan, 1.0000])''')
 
+            y = torch.tensor([4, inf, complex(1.5, inf), complex(-inf, 4), 0, complex(nan, inf), 1])
```
@mruberry here's the test for `NaN` and `inf` cases

mruberry replied:
Can we add a case with nan in the imaginary part?

anjali411 replied:
sure added!

anjali411 committed:
Update on "Fix complex printing for sci_mode=True"


This PR makes the following changes:
1. Complex Printing now uses print formatting for it's real and imaginary values and they are joined at the end.
2. Adding 1. naturally fixes the printing of complex tensors in sci_mode=True

```
>>> torch.tensor(float('inf')+float('inf')*1j)
tensor(nan+infj)
>>> torch.randn(2000, dtype=torch.cfloat)
tensor([ 0.3015-0.2502j, -1.1102+1.2218j, -0.6324+0.0640j,  ...,
        -1.0200-0.2302j,  0.6511-0.1889j, -0.1069+0.1702j])
>>> torch.tensor([1e-3, 3+4j, 1e-5j, 1e-2+3j, 5+1e-6j])
tensor([1.0000e-03+0.0000e+00j, 3.0000e+00+4.0000e+00j, 0.0000e+00+1.0000e-05j,
        1.0000e-02+3.0000e+00j, 5.0000e+00+1.0000e-06j])
>>> torch.randn(3, dtype=torch.cfloat)
tensor([ 1.0992-0.4459j,  1.1073+0.1202j, -0.2177-0.6342j])
>>> x = torch.tensor([1e2, 1e-2])
>>> torch.set_printoptions(sci_mode=False)
>>> x
tensor([  100.0000,     0.0100])
>>> x = torch.tensor([1e2, 1e-2j])
>>> x
tensor([100.+0.0000j,   0.+0.0100j])
```

Differential Revision: [D22309294](https://our.internmc.facebook.com/intern/diff/D22309294)

[ghstack-poisoned]

mruberry reviewed (approved):
LGTM!
